### PR TITLE
[Finishes #101425744] Allow assets to be cached

### PR DESCRIPTION
**Pull request by joshlory**
Currently a `GET` to `/v3/assets/mysUcJO7qONaaF3siN-L3Q/gold_5.png` receives `Cache-Control: private, must-revalidate, max-age=0`.  When an app calls `setImageURL` multiple times a second, each call initiates a new download from the server.

This causes delayed or missing image updates, as image downloads are initiated and canceled in quick succession:

screen shot 2015-08-18 at 5 28 33 pm

Allow caching of assets so the browser doesn't re-download an image each time it is referenced while the app is running.

### CODE REVIEW

File: shared/middleware/files_api.rb [hunk 1]
```diff
@@ -52,7 +52,7 @@ def allowed_file_type?(endpoint, extension)
   # Read a file. Optionally get a specific version instead of the most recent.
   #
   get %r{/v3/(assets|sources)/([^/]+)/([^/]+)$} do |endpoint, encrypted_channel_id, filename|
-    dont_cache
+    dont_cache unless endpoint == 'assets'
```
Comment (philbogle) [inline shared/middleware/files_api.rb]: Could we add tests to make sure that caching behavior is correct for assets and non-assets?

File: apps/src/applab/commands.js [hunk 2]
```diff
@@ -866,6 +874,13 @@ applabCommands.setImageURL = function (opts) {
   var element = document.getElementById(opts.elementId);
   if (divApplab.contains(element) && element.tagName === 'IMG') {
     element.src = Applab.maybeAddAssetPathPrefix(opts.src);
+
+    if (!ensureCached[element.src]) {
+      ensureCached[element.src] = true;
+      var img = new Image();
+      img.src = element.src;
```
Comment (joshlory) [inline apps/src/applab/commands.js]: Is there any risk of this causing a memory leak?  We don't hold a reference to `img` anywhere so it should be eligible for GC as soon as it goes out of scope.

Comment (Bjvanminnen) [inline apps/src/applab/commands.js]: I might move line 879 to be after 881, just to make what's happening a little more obvious (we're loading, then marking as cached).

Also, could we run into any problems where we claim we're cached, but haven't actually loaded the image yet (i.e. after  setting img.src/ensureCached, but before img.onload is called, or if img.onload fails).

**Comment by Bjvanminnen** [comment]
lgtm

File: apps/src/applab/commands.js (hunk 2 above)
Comment (joshlory) [inline apps/src/applab/commands.js]: In this case "cached" is a little confusing.  It just means we've initiated an open-ended request for the asset.  The asset may still be loading.

The problem was that quickly-changing images would keep getting canceled and never make it into the cache.  This ensures that even if `element.src` is changed (which would normally cancel the download), `img.src` will be unchanged and ensure the asset is fully downloaded into the browser's cache.

Comment (Bjvanminnen) [inline apps/src/applab/commands.js]: Maybe ensureImageLoading?

Will we have similar problems for other assets like sounds?

**Pull request merged**